**Проект по SQL**


### ОПИСАНИЕ ПРОЕКТА

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Стартаперы начали создавать приложения для тех, кто любит читать.
Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Первая задача  — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### ОСНОВНАЯ ЦЕЛЬ

- проанализировать базу данных, содержащую информацию о книгах, издательствах, авторах и пользовательских обзорах, чтобы разработать ценностное предложение для нового продукта.

### ЗАДАЧИ

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### ПОСЛЕДОВАТЕЛЬНОСТЬ ВЫПОЛНЕНИЯ ПРОЕКТА

- Загрузка библиотек
- Загрузка данных
- Вывод данных
- Решение задач
- Вывод

### ОПИСАНИЕ ДАННЫХ

Таблица books содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица authors содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

Таблица publishers содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица ratings содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица reviews содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)


In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# Выведем данные таблицы books
books = '''SELECT * FROM books'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(books), con = con)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [5]:
# # Выведем данные таблицы authors
authors = '''SELECT * FROM authors'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(authors), con = con)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [6]:
# # Выведем данные таблицы publishers
publishers = '''SELECT * FROM publishers'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(publishers), con = con)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [7]:
# # Выведем данные таблицы ratings
ratings = '''SELECT * FROM ratings'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(ratings), con = con)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [8]:
# # Выведем данные таблицы reviews
reviews = '''SELECT * FROM reviews'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(reviews), con = con)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [9]:
query = '''SELECT * FROM books WHERE publication_date >= '2000-01-01' '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
816      994        408     World War Z: An Oral History of the Zombie War   
817      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
818      998        201                                    Year of Wonders   
819      999         94                        You Suck (A Love Story  #2)   
820     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
816        342       2006-09-12            68  
817        265       2001-02-06           147  
818        358       2002-04-30           212  
819        328       2007-01-16           331  
820        540       2006-04-25           143  

[821 rows x 6 columns]

**Вывод:**
После 01.01.2000 года вышла 821 книга    

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [10]:
query = '''
SELECT  books.book_id, title, COUNT(distinct review_id) AS review, ROUND(AVG(rating), 2) AS rating
FROM books 
LEFT JOIN reviews ON books.book_id = reviews.book_id 
LEFT JOIN ratings ON books.book_id = ratings.book_id
GROUP BY books.book_id, books.title
ORDER BY review DESC
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  review  \
0        948                            Twilight (Twilight  #1)       7   
1        963                                Water for Elephants       6   
2        734                                   The Glass Castle       6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...       6   
4        695  The Curious Incident of the Dog in the Night-Time       6   
..       ...                                                ...     ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel       0   
996      808                            The Natural Way to Draw       0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites       0   
998      221                          Essential Tales and Poems       0   
999      191  Disney's Beauty and the Beast (A Little Golden...       0   

     rating  
0      3.66  
1      3.98  
2      4.21  
3      4.41  
4      4.08  
..      ...  
995    3.67  
996    3.00  
997    5.00  
998    4.00  
999    4.00  

[1000 rows x 4 columns]

**Вывод:**
По количеству обзоров лидируют "Сумерки"    

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [11]:
query = '''
SELECT publisher, COUNT(book_id) AS sum_books
FROM books 
INNER JOIN publishers ON books.publisher_id = publishers.publisher_id 
WHERE num_pages > 50
GROUP BY publisher
ORDER BY sum_books DESC
LIMIT 1
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  sum_books
0  Penguin Books         42

**Вывод:**
Больше всего книг толще 50 страниц выпустило издательство Penguin Books     

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [12]:
query = '''
SELECT author, ROUND(AVG(rating), 2) AS rating
FROM authors 
INNER JOIN books ON books.author_id = authors.author_id 
INNER JOIN ratings ON ratings.book_id = books.book_id 
WHERE authors.author_id IN 
(
    SELECT authors.author_id
    FROM authors
    INNER JOIN books ON authors.author_id = books.author_id
    INNER JOIN ratings ON books.book_id = ratings.book_id
    GROUP BY authors.author_id
    HAVING COUNT(ratings.rating) > 50
)
GROUP BY author
ORDER BY rating DESC
LIMIT 1
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  rating
0  J.K. Rowling/Mary GrandPré    4.29

**Вывод:**
Самый высокий рейтинг у Джоан Роулинг

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [13]:
query = '''
WITH 
USERS AS 
(SELECT username AS USER FROM ratings 
GROUP BY username
HAVING COUNT(ratings.rating) > 48),

REVIEW AS
(SELECT reviews.username AS USER, count(reviews.review_id) AS CNT_review
FROM reviews
GROUP BY reviews.username
ORDER BY CNT_review DESC)

SELECT AVG(CNT_review) 
FROM REVIEW 
INNER JOIN USERS ON REVIEW.USER=USERS.USER

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

**Вывод:**
Среднее количеств обзоров от авторов, которые поставили много оценок - 24 обзора    

**ВЫВОД**


**Задачи и результаты** 

- Было выпущено 821 книга после 1 января 2000 года. Это значительное количество, указывающее на активное производство книг в последние десятилетия.

- Среди книг, наибольшее количество обзоров собрала книга "Сумерки". Это свидетельствует о высокой популярности и обсуждаемости данной книги среди читателей.

- Издательство Penguin Books выпустило больше всего книг, которые толще 50 страниц. Это указывает на высокую активность издательства в производстве полноценных книг, исключая брошюры.

- Джоан Роулинг оказалась автором с самой высокой средней оценкой книг, учитывая только те книги, которые получили 50 и более оценок. Это подчеркивает признание и популярность её работ среди большого числа читателей.

- Среднее количество обзоров от активных пользователей (которые оставили более 48 оценок) составляет 24 обзора. Это говорит о высоком уровне вовлеченности и активности данной группы пользователей.

**Заключение**

Анализ базы данных показал, что рынок книг после 2000 года является активным и разнообразным. Издательство Penguin Books лидирует по количеству полноценных книг, а Джоан Роулинг имеет наивысший средний рейтинг среди авторов. Книга "Сумерки" является самой обсуждаемой, а активные пользователи показывают высокую вовлеченность, оставляя множество отзывов. Эти данные могут помочь в формировании ценностного предложения для нового продукта, ориентированного на активных читателей и популярные жанры и авторов.